File: output_repair/claude-opus/sample_3/repair_3/c/c/Struct/account.c
#include "stdlib.h"

struct account
{
    int balance;
};

/*@ predicate account(struct account *a) = a->balance |-> _; @*/

struct account *create_account()
//@ requires true;
//@ ensures malloc_block_account(result) &*& result->balance |-> _ &*& result != 0;
{
    struct account *myAccount = malloc(sizeof(struct account));
    if (myAccount == 0)
    {
        abort();
    }
    myAccount->balance = 0;
    return myAccount;
}

void account_set_balance(struct account *myAccount, int newBalance)
//@ requires myAccount->balance |-> _;
//@ ensures myAccount->balance |-> _;
{
    myAccount->balance = newBalance;
}

void account_dispose(struct account *myAccount)
//@ requires malloc_block_account(myAccount) &*& myAccount->balance |-> _;
//@ ensures true;
{
    free(myAccount);
}

int main()
//@ requires true;
//@ ensures true;
{
    struct account *myAccount = create_account();
    account_set_balance(myAccount, 5);
    account_dispose(myAccount);
    return 0;
}

File: output/qwen3/rust/purely_unsafe/solutions/account.rs
use std::alloc::{alloc, dealloc, handle_alloc_error, Layout};

#[pred account(p: *mut Account) = 
    p != 0 &*& 
    malloc_block_Account(p, _) &*& 
    (*p).balance |-> ?b
]

struct Account {
    balance: i32,
}

impl Account {
    #[requires(Layout::new::<Account>().size() > 0)]
    #[ensures(account(result))]
    unsafe fn create() -> *mut Account {
        let my_account = alloc(Layout::new::<Account>()) as *mut Account;
        if my_account.is_null() {
            handle_alloc_error(Layout::new::<Account>());
        }
        (*my_account).balance = 0;
        my_account
    }

    #[requires(account(my_account) &*& new_balance |-> ?nb)]
    #[ensures(account(my_account))]
    unsafe fn set_balance(my_account: *mut Account, new_balance: i32) {
        (*my_account).balance = new_balance;
    }

    #[requires(account(my_account))]
    #[ensures(true)]
    unsafe fn dispose(my_account: *mut Account) {
        dealloc(my_account as *mut u8, Layout::new::<Account>());
    }
}

fn main() {
    unsafe {
        let my_account = Account::create();
        Account::set_balance(my_account, 5);
        Account::dispose(my_account);
    }
}